### ISSUE TITLE
class RequestUtil could not be found after upgrading to Vaadin 20

### ISSUE THREAD

**Issue by Avec112** [vaadin/spring#848]
So I tried to upgrade from Vaadin 19.0.8 to 20.0.1 but no luck.

The Spring Container failed to start and returned this description.

_Field requestUtil in `com.caadin.flow.spring.security.VaadinDefaultRequestCache` required a bean of type '`com.vaadin.flow.spring.security.RequestUtil`' that could not be found._

I do have `vaadin-spring:17.0.0` in classpath and `RequestUtil` (annotated with @Component) exists.

I have changed my code from extending `WebSecurityConfigurerAdapter` to `VaadinWebSecurityConfigurerAdapter` but I get the same error regardless. I followed this upgrade [documentation](https://vaadin.com/docs/latest/guide/upgrading/#default-spring-security-configuration). 

Am I missing something or is something wrong here?

**Comment by Artur-** [comment]
Are you using plain Spring or Spring Boot? For Spring Boot the bean is registered in `SpringBootAutoConfiguration`

**Comment by Avec112** [comment]
I am using Vaadin Flow by the way. Maybe it is not ready for Flow only Fusion?

**Comment by Avec112** [comment]
> Are you using plain Spring or Spring Boot? For Spring Boot the bean is registered in `SpringBootAutoConfiguration`

I am using Spring Boot.

**Comment by Avec112** [comment]
So I am still struggling  with how I can be able to upgrade without getting the error mentioned above. 
Any help would be appreciated.

**Comment by Artur-** [comment]
If you can publish a project where the problem reproduces, then somebody else would be able to take a look.

Otherwise, you could grab a project from start.vaadin.com and try to compare and find differences

**Comment by Avec112** [comment]
Okay thanks. I hoped this was a known possible issue and someone knew a solution for it.

It is impossible to provide our project unfortunately. We do all our work on a offline LAN. The distance from start.vaadin.com applications gets bigger and bigger over time meaning our application is getting more and more complicated compared to a "default start application". The upgrades went without to much problems from 14 and up until now. At least there was nothing in the documentation that pointed out anything special.

I guess I can start with a 19.0.8 project and try to upgrade it and hope the same problem occur. Alternatively I will have to compare for differences as you mentioned.

**Comment by denis-anisimov** [comment]
Sorry, but until you can provide the code that reproduces the issue, there is nothing that we can do. Until that we consider this an issue specific to your application. Please reopen the issue if you discover a root cause not dependent on your application and can provide the code to reproduce the issue.

**Comment by Avec112** [comment]
So I am back from vacation this week and I just took another look at this issue and I think I have found a temporary fix for my problem but the main fix must be done in the library it self. 

TL;DR: Drop the use field injection by use of '@Autowired` and always use constructor injection and at the same time make the fields `final`. This will not lead into null since it will force any dependency bean to be instantiated at the right time.  


The problem is that bean `RequestUtil` does not exist in the container when bean `vaadinDefaultRequestCache` starts using it. I was able to circumvent the problem by doing a little hack. I created my own class `MyRequestUtil extends RequestUtil` and made it an `@Component` and no implementation (empty class). After this change I was able to successfully load the container and run the application. 

Spring Container can sometimes be mysterious at best regarding bean instantiation. I have experienced random bean loading many times and this is something we as developers should take into account by forcing the correct behavior. In project I used to work on me and my colleague could have the exact same code but different outcome regarding bean loading in the container. It could work for me while he could experience a circular dependency. Other days I could go compile and run the application successfully and go home. The next day I could come back to non starting application because of a new circular dependency. The circular problem is not the point here but the random bean loading is. 

The code in `Vaadin Spring` heavily rely on the use of `@Autowired` (field injection. Actually reflection with setter injection and can therefor be null). Constructor injection is highly recommended and will at the same time ensure that any dependency (beans) is created upfront (no nulls). The field should also become `final`. Constructor injection over field injection has been known to be the recommendation for many years in the Spring Community. 

Some documentation:
* https://medium.com/engineering-zemoso/when-not-to-autowire-in-spring-spring-boot-93e6a01cb793
* https://www.logicbig.com/tutorials/spring-framework/spring-core/types-of-dependency-injection.html
* https://odrotbohm.de/2013/11/why-field-injection-is-evil/ 
* https://www.codecleaner.org/autowired-is-evil/

What Pivotal says:

> **Constructor-based or setter-based DI?**
> 
> Since you can mix constructor-based and setter-based DI, it is a good rule of thumb to use constructors for mandatory dependencies and setter methods or configuration methods for optional dependencies. Note that use of the @Required annotation on a setter method can be used to make the property a required dependency.
> 
> **The Spring team generally advocates constructor injection as it enables one to implement application components as immutable objects and to ensure that required dependencies are not null.** Furthermore constructor-injected components are always returned to client (calling) code in a fully initialized state. As a side note, a large number of constructor arguments is a bad code smell, implying that the class likely has too many responsibilities and should be refactored to better address proper separation of concerns.
> 
> Setter injection should primarily only be used for optional dependencies that can be assigned reasonable default values within the class. Otherwise, not-null checks must be performed everywhere the code uses the dependency. One benefit of setter injection is that setter methods make objects of that class amenable to reconfiguration or re-injection later. Management through JMX MBeans is therefore a compelling use case for setter injection.
> 
> Use the DI style that makes the most sense for a particular class. Sometimes, when dealing with third-party classes for which you do not have the source, the choice is made for you. For example, if a third-party class does not expose any setter methods, then constructor injection may be the only available form of DI.

I cannot be 100% sure this is the problem, but I would bet my two cents that it is. Why your code does not end with a null bean and I do I cannot explain but I believe it is because of bean instantiation randomness. Using constructor injection will force the bean to be created at the right time.

**Comment by Avec112** [comment]
@denis-anisimov I cannot open this issue again as far as I can see.

**Comment by denis-anisimov** [comment]
No problem , please publish a project which can be used to reproduce an issue as requested here https://github.com/vaadin/spring/issues/848#issuecomment-865625360 and I will reopen the ticket.

**Comment by Avec112** [comment]
As I have explained earlier it is impossible to share our code (security reasons...not even connected to Internet). Also as I just explained the `Vaadin Spring` library should implement dependencies with **constructor injection** not field injection (@Autowired uses setter injection by reflection as explained and is causing the problem). I would not experience a missing dependency if constructor injection is implemented. This has nothing to do with my code since the implementation is the problem. I think I explained this extensively in my last post.

**Comment by denis-anisimov** [comment]
Sorry, I'm just a dumb issue reports checker. 
I'm not familiar with the functionality and I'm not a developer who is going to fix this.

My goal is just to check the issue report: the steps to reproduce are mandatory. The suggestion how to fix and why it happens is optional.

>As I have explained earlier it is impossible to share our code (security reasons...not even connected to Internet).

That's totally fine. There is no need in the real project. Please use this project https://github.com/vaadin/skeleton-starter-flow-spring as a basis and make a simple project which can be used to see an issue.
The extra code from the real project contains too many irrelevant details which makes harder to find the real reason.

>Also as I just explained the Vaadin Spring library should implement dependencies with constructor injection not field injection (@Autowired uses setter injection by reflection as explained and is causing the problem).

I would like to see the official Spring documentation which says that.
As far as I know the most of the core Spring add-on functionality is written using `@Autowired ` for fields and there were no any complains about this.
There are only two cases :`@Autowired`  _may_ be used for fields or it _may not be_. There is no such thing as "shouldn't".
The annotation has `@Target` `ElementType.FIELD` which  defines a contract: it _may_ be use for a field.
There is only one limitation related to using `@Autowired`  with a field: you may not use the value in the constructor since it's not initialized until the object is created.
So if `@Autowired`  field injection doesn't work then it's a bug in Spring.

There is no any difference in using  `@Autowired` for a field or for the constructor parameter (unless there is a bug in Spring).
It's just a code style preference and best practices question.

What does matter : the order of bean creation and race condition. It looks like in this case it's a question of race condition.
For some reasons `RequestUtil` bean is not yet created (or even registered) by Spring when it's used inside `VaadinDefaultRequestCache`. And since it's not used in the constructor it doesn't matter whether it's a field injection or constructor.

> This has nothing to do with my code since the implementation in Vaadin Spring seems to be the problem. I think I explained this extensively in my last post.

It's not possible to say what causes the problem until there is a way to see the problem. It might be that there is a genuine bug in Vaadin Spring add-on. That's quite possible. But we should be able to identify it using steps to reproduce.
And having the steps to reproduce is the one way to prove that the bug is fixed: the steps to reproduce without a fix leads to the incorrect behavior and with the fix the behavior should be correct. 

Anyway: as I said I'm not a developer of this functionality and can't go into details of this issue since I don't know anything about this. I cannot reopen the ticket if there is no steps to reproduce or attached project which we can use to reproduce.

**Comment by Avec112** [comment]
Thanks a lot.

You are absolutely right. It must have something to do with race condition. In my case the `Vaadin Spring` bean `requestUtil` is not in place when it is needed. Since this is a `Vaadin Spring` internal class I cannot see how my code ever could be the problem unless I have a class with the same name instantiated before yours .  As I wrote I am able to "fix" it by extending it and making my own dumb version. I will take a look on the code when I am back Thursday. Maybe there is a bean naming conflict somewhere from another spring library. I will take a look asap. 

Some older official Spring documentation regarding dependency injection recommendations and when to use what:
* https://spring.io/blog/2007/07/11/setter-injection-versus-constructor-injection-and-the-use-of-required/
* https://docs.spring.io/spring-framework/docs/3.0.0.M4/reference/html/ch03s04.html (see 3.4.1.3)
* https://docs.spring.io/spring-framework/docs/4.2.x/spring-framework-reference/html/beans.html#beans-constructor-injection (bottom)

**Comment by Avec112** [comment]
I see I believed field injection later used reflection to call setter injection. That is not correct. Field injection with `@Autowired` uses reflection and sets value directly to that field. 

Her is another great article about the different ways to do DI and what is recommended. 
https://www.vojtechruzicka.com/field-dependency-injection-considered-harmful/

**Comment by Avec112** [comment]
I have found the problem on my side. 

After the introduction of `Vaadin Spring` (upgrade from Vaadin 19 to 20) one of my tests failed with `UnsatisfiedDependencyException` because bean `requestUtil` was not found. I ran a stripped down `@SpringBootTest`. The particular test ran fine before upgrade so that was what threw me of. I have now included class `RequestUtil` and that solved the problem.

```
@SpringBootTest (
  classes = { Application.class, RequestUtil.class }, // had to add RequestUtil for it to work
  webEnvironment = SpringBootTest.WebEnvironment.NONE // for faster load

)
```

I will still recommend constructor injection however, but I admit that had nothing to do with this particular problem. This issue can therefor remain closed.
